File: Plugins/AM_Mz_Refinery_Plugin/clsMzRefineryMassErrorStatsExtractor.vb
Imports AnalysisManagerBase
Imports System.IO
Imports System.Text
Imports System.Runtime.InteropServices

''' <summary>
''' This class reads the console text from the PPMErrorCharter's console output and extracts the parent ion mass error information
''' It passes on the information to DMS for storage in table T_Dataset_QC
''' </summary>
''' <remarks></remarks>
Public Class clsMzRefineryMassErrorStatsExtractor

    Protected Const STORE_MASS_ERROR_STATS_SP_NAME As String = "StoreDTARefMassErrorStats"

    Protected m_mgrParams As IMgrParams
    Protected m_WorkDir As String
    Protected m_DebugLevel As Short
    Protected mPostResultsToDB As Boolean

    Protected mErrorMessage As String

    Protected Structure udtMassErrorInfoType
        Public DatasetName As String
        Public DatasetID As Integer
        Public PSMJob As Integer
        Public MassErrorPPM As Double				' Parent Ion Mass Error, before refinement
        Public MassErrorPPMRefined As Double		' Parent Ion Mass Error, after refinement
    End Structure

    Public ReadOnly Property ErrorMessage() As String
        Get
            Return mErrorMessage
        End Get
    End Property

    Public Sub New(ByRef mgrParams As IMgrParams, strWorkDir As String, intDebugLevel As Short, blnPostResultsToDB As Boolean)

        m_mgrParams = mgrParams
        m_WorkDir = strWorkDir
        m_DebugLevel = intDebugLevel
        mPostResultsToDB = blnPostResultsToDB

        mErrorMessage = String.Empty
    End Sub

    Protected Function ConstructXML(udtMassErrorInfo As udtMassErrorInfoType) As String
        Dim sbXml = New StringBuilder()

        Try
            sbXml.Append("<DTARef_MassErrorStats>")

            sbXml.Append((Convert.ToString("<Dataset>") & udtMassErrorInfo.DatasetName) + "</Dataset>")
            sbXml.Append((Convert.ToString("<PSM_Source_Job>") & udtMassErrorInfo.PSMJob) + "</PSM_Source_Job>")

            sbXml.Append("<Measurements>")
            sbXml.Append((Convert.ToString("<Measurement Name=""" + "MassErrorPPM" + """>") & udtMassErrorInfo.MassErrorPPM) + "</Measurement>")
            sbXml.Append((Convert.ToString("<Measurement Name=""" + "MassErrorPPM_Refined" + """>") & udtMassErrorInfo.MassErrorPPMRefined) + "</Measurement>")
            sbXml.Append("</Measurements>")

            sbXml.Append("</DTARef_MassErrorStats>")

        Catch ex As Exception
            Console.WriteLine("Error converting Mass Error stats to XML; details:")
            Console.WriteLine(ex)
            Return String.Empty
        End Try

        Return sbXml.ToString()

    End Function

    Public Function ParsePPMErrorCharterOutput(
      strDatasetName As String,
      intDatasetID As Integer,
      intPSMJob As Integer,
      ppmErrorCharterConsoleOutputFilePath As String) As Boolean

        ' Parse the Console Output file to extract the mass error reported in this table
        '
        ' Using fixed data file "E:\DMS_WorkDir\Pcarb001_LTQFT_run1_23Sep05_Andro_0705-06_FIXED.mzML"
        ' Statistic                   Original    Refined
        ' MeanMassErrorPPM:              2.430      1.361
        ' MedianMassErrorPPM:            1.782      0.704
        ' StDev(Mean):                   6.969      6.972
        ' StDev(Median):                 6.999      7.003
        ' PPM Window for 99%: 0 +/-     22.779     21.712
        ' PPM Window for 99%: high:     22.779     21.712
        ' PPM Window for 99%:  low:    -19.216    -20.305

        Const MASS_ERROR_PPM As String = "MedianMassErrorPPM:"

        Try

            Dim udtMassErrorInfo = New udtMassErrorInfoType
            udtMassErrorInfo.DatasetName = strDatasetName
            udtMassErrorInfo.DatasetID = intDatasetID
            udtMassErrorInfo.PSMJob = intPSMJob
            udtMassErrorInfo.MassErrorPPM = Double.MinValue
            udtMassErrorInfo.MassErrorPPMRefined = Double.MinValue

            Dim fiSourceFile = New FileInfo(ppmErrorCharterConsoleOutputFilePath)
            If Not fiSourceFile.Exists Then
                mErrorMessage = "MzRefinery Log file not found"
                Return False
            End If

            Using srSourceFile = New StreamReader(New FileStream(fiSourceFile.FullName, FileMode.Open, FileAccess.Read, FileShare.Read))

                Do While Not srSourceFile.EndOfStream
                    Dim strLineIn = srSourceFile.ReadLine()

                    If String.IsNullOrWhiteSpace(strLineIn) Then Continue Do

                    strLineIn = strLineIn.Trim()

                    Dim massError As Double

                    If strLineIn.StartsWith(MASS_ERROR_PPM) Then
                        Dim dataString = strLineIn.Substring(MASS_ERROR_PPM.Length).Trim()
                        massError = 0

                        Dim dataValues = dataString.Split(" "c).ToList()

                        If Double.TryParse(dataValues.First, massError) Then
                            udtMassErrorInfo.MassErrorPPM = massError
                        End If

                        If dataValues.Count > 1 AndAlso Double.TryParse(dataValues.Last, massError) Then
                            udtMassErrorInfo.MassErrorPPMRefined = massError
                        End If

                    End If
                Loop

            End Using

            If Math.Abs(udtMassErrorInfo.MassErrorPPM - Double.MinValue) < Single.Epsilon Then
                mErrorMessage = "Did not find '" & MASS_ERROR_PPM & "' in the PPM Error Charter output"
                Return False
            End If

            If Math.Abs(udtMassErrorInfo.MassErrorPPMRefined - Double.MinValue) < Single.Epsilon Then
                mErrorMessage = "Did not find '" & MASS_ERROR_PPM & "' with two values in the PPM Error Charter output"
                Return False
            End If

            Dim strXMLResults = ConstructXML(udtMassErrorInfo)

            If mPostResultsToDB Then
                Dim blnSuccess As Boolean

                blnSuccess = PostMassErrorInfoToDB(intDatasetID, strXMLResults)

                If Not blnSuccess Then
                    If String.IsNullOrEmpty(mErrorMessage) Then
                        mErrorMessage = "Unknown error posting Mass Error results from MzRefinery to the database"
                    End If
                    Return False
                End If
            End If

        Catch ex As Exception
            mErrorMessage = "Exception in ParsePPMErrorCharterOutput: " & ex.Message
            Return False
        End Try

        Return True

    End Function

    Protected Function PostMassErrorInfoToDB(
      intDatasetID As Integer,
      strXMLResults As String) As Boolean

        Const MAX_RETRY_COUNT As Integer = 3

        Dim objCommand As System.Data.SqlClient.SqlCommand

        Dim blnSuccess As Boolean

        Try

            ' Call stored procedure StoreDTARefMassErrorStats in DMS5

            objCommand = New System.Data.SqlClient.SqlCommand()

            With objCommand
                .CommandType = CommandType.StoredProcedure
                .CommandText = STORE_MASS_ERROR_STATS_SP_NAME

                .Parameters.Add(New SqlClient.SqlParameter("@Return", SqlDbType.Int))
                .Parameters.Item("@Return").Direction = ParameterDirection.ReturnValue

                .Parameters.Add(New SqlClient.SqlParameter("@DatasetID", SqlDbType.Int))
                .Parameters.Item("@DatasetID").Direction = ParameterDirection.Input
                .Parameters.Item("@DatasetID").Value = intDatasetID

                .Parameters.Add(New SqlClient.SqlParameter("@ResultsXML", SqlDbType.Xml))
                .Parameters.Item("@ResultsXML").Direction = ParameterDirection.Input
                .Parameters.Item("@ResultsXML").Value = strXMLResults
            End With


            Dim objAnalysisTask = New clsAnalysisJob(m_mgrParams, m_DebugLevel)

            'Execute the SP (retry the call up to 4 times)
            Dim ResCode As Integer
            ResCode = objAnalysisTask.DMSProcedureExecutor.ExecuteSP(objCommand, MAX_RETRY_COUNT)

            objAnalysisTask = Nothing

            If ResCode = 0 Then
                blnSuccess = True
            Else
                mErrorMessage = "Error storing MzRefinery Mass Error Results in the database, " & STORE_MASS_ERROR_STATS_SP_NAME & " returned " & ResCode.ToString
                blnSuccess = False
            End If

        Catch ex As System.Exception
            mErrorMessage = "Exception storing MzRefinery Mass Error Results in the database: " & ex.Message
            blnSuccess = False
        End Try

        Return blnSuccess

    End Function

End Class

File: Plugins/AM_Mz_Refinery_Plugin/clsMzRefineryMassErrorStatsExtractor.cs
using AnalysisManagerBase;
using System;
using System.Data;
using System.IO;
using System.Linq;
using System.Text;
using PRISMDatabaseUtils;

namespace AnalysisManagerMzRefineryPlugIn
{
    /// <summary>
    /// This class reads the console text from the PPMErrorCharter's console output and extracts the parent ion mass error information
    /// It passes on the information to DMS for storage in table T_Dataset_QC
    /// </summary>
    public class clsMzRefineryMassErrorStatsExtractor
    {
        private const string STORE_MASS_ERROR_STATS_SP_NAME = "StoreDTARefMassErrorStats";
        private readonly IMgrParams mMgrParams;
        private readonly short mDebugLevel;

        private readonly bool mPostResultsToDB;

        public string ErrorMessage { get; private set; }

        /// <summary>
        /// Mass error info; populated by ParsePPMErrorCharterOutput
        /// </summary>
        public clsMassErrorInfo MassErrorInfo { get; }

        /// <summary>
        /// Constructor
        /// </summary>
        /// <param name="mgrParams"></param>
        /// <param name="debugLevel"></param>
        /// <param name="postResultsToDB"></param>
        public clsMzRefineryMassErrorStatsExtractor(IMgrParams mgrParams, short debugLevel, bool postResultsToDB = true)
        {
            mMgrParams = mgrParams;
            mDebugLevel = debugLevel;
            mPostResultsToDB = postResultsToDB;

            ErrorMessage = string.Empty;

            MassErrorInfo = new clsMassErrorInfo();
            MassErrorInfo.Clear();
        }

        private string ConstructXML()
        {
            var sbXml = new StringBuilder();

            try
            {
                sbXml.Append("<DTARef_MassErrorStats>");

                sbXml.Append(Convert.ToString("<Dataset>") + MassErrorInfo.DatasetName + "</Dataset>");
                sbXml.Append(Convert.ToString("<PSM_Source_Job>") + MassErrorInfo.PSMJob + "</PSM_Source_Job>");

                sbXml.Append("<Measurements>");
                sbXml.Append(Convert.ToString("<Measurement Name=\"" + "MassErrorPPM" + "\">") + MassErrorInfo.MassErrorPPM + "</Measurement>");
                sbXml.Append(Convert.ToString("<Measurement Name=\"" + "MassErrorPPM_Refined" + "\">") + MassErrorInfo.MassErrorPPMRefined + "</Measurement>");
                sbXml.Append("</Measurements>");

                sbXml.Append("</DTARef_MassErrorStats>");
            }
            catch (Exception ex)
            {
                Console.WriteLine("Error converting Mass Error stats to XML; details:");
                Console.WriteLine(ex);
                return string.Empty;
            }

            return sbXml.ToString();
        }

        /// <summary>
        /// Parse the PPM Error Charter console output file to extract the mass error reported in this table
        /// </summary>
        /// <param name="datasetName"></param>
        /// <param name="datasetID"></param>
        /// <param name="psmJob"></param>
        /// <param name="ppmErrorCharterConsoleOutputFilePath"></param>
        public bool ParsePPMErrorCharterOutput(string datasetName, int datasetID, int psmJob, string ppmErrorCharterConsoleOutputFilePath)
        {
            // Example console output:
            //
            // Using fixed data file "E:\DMS_WorkDir\DatasetName_FIXED.mzML"
            // Statistic                   Original    Refined
            // MeanMassErrorPPM:              2.430      1.361
            // MedianMassErrorPPM:            1.782      0.704
            // StDev(Mean):                   6.969      6.972
            // StDev(Median):                 6.999      7.003
            // PPM Window for 99%: 0 +/-     22.779     21.712
            // PPM Window for 99%: high:     22.779     21.712
            // PPM Window for 99%:  low:    -19.216    -20.305

            const string MASS_ERROR_PPM = "MedianMassErrorPPM:";

            try
            {
                MassErrorInfo.Clear();
                MassErrorInfo.DatasetName = datasetName;
                MassErrorInfo.PSMJob = psmJob;

                var sourceFile = new FileInfo(ppmErrorCharterConsoleOutputFilePath);
                if (!sourceFile.Exists)
                {
                    ErrorMessage = "MzRefinery Log file not found";
                    return false;
                }

                using (var reader = new StreamReader(new FileStream(sourceFile.FullName, FileMode.Open, FileAccess.Read, FileShare.Read)))
                {
                    while (!reader.EndOfStream)
                    {
                        var dataLine = reader.ReadLine();

                        if (string.IsNullOrWhiteSpace(dataLine))
                            continue;

                        dataLine = dataLine.Trim();

                        if (!dataLine.StartsWith(MASS_ERROR_PPM))
                            continue;

                        var dataString = dataLine.Substring(MASS_ERROR_PPM.Length).Trim();

                        var dataValues = dataString.Split(' ').ToList();

                        if (double.TryParse(dataValues.First(), out var massError))
                        {
                            MassErrorInfo.MassErrorPPM = massError;
                        }

                        if (dataValues.Count > 1 && double.TryParse(dataValues.Last(), out massError))
                        {
                            MassErrorInfo.MassErrorPPMRefined = massError;
                        }
                    }
                }

                if (Math.Abs(MassErrorInfo.MassErrorPPM - double.MinValue) < float.Epsilon)
                {
                    // Did not find 'MedianMassErrorPPM' in the PPM Error Charter output
                    ErrorMessage = "Did not find '" + MASS_ERROR_PPM + "' in the PPM Error Charter output";
                    return false;
                }

                if (Math.Abs(MassErrorInfo.MassErrorPPMRefined - double.MinValue) < float.Epsilon)
                {
                    // Did not find 'MedianMassErrorPPM' with two values in the PPM Error Charter output
                    ErrorMessage = "Did not find '" + MASS_ERROR_PPM + "' with two values in the PPM Error Charter output";
                    return false;
                }

                var xmlResults = ConstructXML();

                if (mPostResultsToDB)
                {
                    var success = PostMassErrorInfoToDB(datasetID, xmlResults);

                    if (!success)
                    {
                        if (string.IsNullOrEmpty(ErrorMessage))
                        {
                            ErrorMessage = "Unknown error posting Mass Error results from MzRefinery to the database";
                        }
                        return false;
                    }
                }
            }
            catch (Exception ex)
            {
                ErrorMessage = "Exception in ParsePPMErrorCharterOutput: " + ex.Message;
                return false;
            }

            return true;
        }

        private bool PostMassErrorInfoToDB(int datasetID, string xmlResults)
        {
            const int MAX_RETRY_COUNT = 3;

            bool success;

            try
            {
                var analysisTask = new clsAnalysisJob(mMgrParams, mDebugLevel);
                var dbTools = analysisTask.DMSProcedureExecutor;

                // Call stored procedure StoreDTARefMassErrorStats in DMS5
                // Data is stored in table T_Dataset_QC
                var sqlCmd = dbTools.CreateCommand(STORE_MASS_ERROR_STATS_SP_NAME, CommandType.StoredProcedure);

                dbTools.AddParameter(sqlCmd, "@Return", SqlType.Int, ParameterDirection.ReturnValue);
                dbTools.AddTypedParameter(sqlCmd, "@DatasetID", SqlType.Int, value: datasetID);
                dbTools.AddParameter(sqlCmd, "@ResultsXML", SqlType.XML).Value = xmlResults;

                // Execute the SP (retry the call up to 3 times)
                var resCode = dbTools.ExecuteSP(sqlCmd, MAX_RETRY_COUNT);

                if (resCode == 0)
                {
                    success = true;
                }
                else
                {
                    ErrorMessage = string.Format(
                        "Error storing MzRefinery Mass Error Results in the database, {0} returned {1}",
                        STORE_MASS_ERROR_STATS_SP_NAME, resCode);
                    success = false;
                }
            }
            catch (Exception ex)
            {
                ErrorMessage = "Exception storing MzRefinery Mass Error Results in the database: " + ex.Message;
                success = false;
            }

            return success;
        }
    }
}
